### PR TITLE
fix pyfile and other language specific dependency across different jb…

### PR DESCRIPTION
**Pull request by Will-ShaoHua**
## Problems

Original PR #3800 was supposed to make all implementation of `CodeWhispererFileCrawler` and corresponding `CodeWhispererProgrammingLanguage` live in its IDE module, e.g. JS related languages/filecrawler should go in `Ultimate` however it will change the current behavior of CodeWhisperer. Current status quo is even on the IDE itself is not able to recognize the file type, for example, `.js` `.ts` file on `Intellij Community`, users are still able to trigger CodeWhisperer service on `.js` `.ts` files as CodeWhisperer is based on the file extension instead of how IDE read such files. By making `CodeWhispererProgrammingLanguage` and `CodeWhispererFileCrawler` extension point which will change the behavior and users will NOT be able to trigger CodeWhisperer service if the IDE doesn't recognize the file type.

## Before

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/1d88f2ec-f484-49b9-9ee1-6b7c116f7bcb

## After

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/18a6b6d9-fd90-49da-a08c-1d63d5c5a3b9

## Solution
We extract the logic which deal with reading the psi elements and make it an extension point.

## Changes
- Extract `FileCrawler.findSourceFileByContent` to an extension point interface because the implementation is leveraging JB's language specific API such as `PyFile`, `ClassOwner` which only exist in the class path for specific IDEs or when specific JB plugin is installed. For example, `Pyfile` only exist when JB python plugin is installed or it's on Pycharm.

### Crossfile 
- should fetch same file extension or belongs target's dialects (ts vs. tsx) 
- should exclude test file
  - [x] Javascript & JSX
  - [x] Typescript & TSX
  - [x] Python
  - [x] Java
### UTG
  - [x] Python
  - [x] Java  

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/JavaCodeWhispererFileCrawler.kt [hunk 1]
```diff
@@ -83,17 +85,13 @@ object JavaCodeWhispererFileCrawler : CodeWhispererFileCrawler() {
...
-        }
-
-        val classAndMethod = runReadAction {
-            myPsiFile.classes.mapNotNull { clazz ->
-                listOfNotNull(clazz.name) + clazz.methods.mapNotNull { method -> method.name }
-            }.flatten()
-        }
+    override fun findSourceFileByContent(target: PsiFile): VirtualFile? = searchRelevantFileInEditors(target) { myPsiFile ->
+        CodeWhispererClassResolver.EP_NAME.findFirstSafe { it is CodeWhispereJavaClassResolver }?.let {
+            val classAndMethos = it.resolveClassAndMembers(myPsiFile)
```
Comment (andrewyuq) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/JavaCodeWhispererFileCrawler.kt:90]: typo

**Review by andrewyuq — APPROVED** [review]

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/JavaCodeWhispererFileCrawler.kt (hunk 1 above)
Comment (Will-ShaoHua) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/JavaCodeWhispererFileCrawler.kt:90]: gotcha

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/CodeWhispererFileCrawlerTest.kt [hunk 2]
```diff
@@ -22,31 +22,115 @@ import software.aws.toolkits.jetbrains.services.codewhisperer.util.JavaCodeWhisp
...
+    fun `searchKeywordsInOpenedFile is language agnostic`() {
+        sut = JavaCodeWhispererFileCrawler
+
+        val targetFile = fixture.addFileToProject("Foo.ts", "I have 10 Foo in total, Foo, Foo, Foo, Foo, Foo, Foo, Foo, Foo, Foo")
+
+        val file0 = fixture.addFileToProject("file0.java", "I have 7 Foo, Foo, Foo, Foo, Foo, Foo, Foo, but I am a pyfile")
+        val file1 = fixture.addFileToProject("File1.ts", "I have 4 Foo key words : Foo, Foo, Foo")
+        val file2 = fixture.addFileToProject("File2.ts", "I have 2 Foo Foo")
+        val file3 = fixture.addFileToProject("File3.ts", "I have only 1 Foo")
+        val file4 = fixture.addFileToProject("File4.ts", "bar bar bar, i have a lot of bar")
```
Comment (rli) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/CodeWhispererFileCrawlerTest.kt:91]: i recommend documenting the expectation that:
```
on Intellij Community, users are still able to trigger CodeWhisperer service on .js .ts files 
```

on this test so we dont forget

**Review by rli — APPROVED** [review]

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/CodeWhispererFileCrawlerTest.kt (hunk 2 above)
Comment (Will-ShaoHua) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/CodeWhispererFileCrawlerTest.kt:91]: sg

**Pull request merged**